Corrections to table selection pasting improvements PR

mlewand opened the pull request:
## What is the purpose of this pull request?

Bug fix

 ## Does your PR contain necessary tests?

- [x] Unit tests
- [x] Manual tests (unit tests inherited from intiial PR)

 ## What changes did you make?

Since paste listener was huge, and very difficult to understand I had to split the logic into smaller pieces, so that it's easier to work with it.

The biggest change is encapsulated `TableSelection` type in tableselection plugin, so that there's no `firstCell`, `lastRow` etc reassigning all over the place. Instead new types take care of that. It also features methods related to the manipulation on the table selection itself.

Also `insertRow` and `insertColumn` in tabletools now returns added DOM elements.

Comandeer commented on plugins/tableselection/plugin.js:
```diff
@@ -538,57 +754,43 @@
 			// cells must be emptied after pasting (#520).
 			editor.once( 'afterPaste', function() {
-				emptyCells( selectedCells.slice( 1 ), true );
+				emptyCells( tableSel.cells.all.slice( 1 ), true );
```
Maybe we could move `emptyCells` into `TableSelection.prototype`?

mlewand replied:
Did it although emptyCells no longer features undo logic (it's been inlined in host code) as it's only used in that one single place.

Comandeer commented on plugins/tabletools/plugin.js:
```diff
@@ -146,6 +147,8 @@
 
 		insertBefore ? newRow.insertBefore( row ) : newRow.insertAfter( row );
+
+		return newRow;
```
There are no tests for these new return values. It would be nice to add them.

mlewand commented on plugins/tableselection/plugin.js:
```diff
@@ -443,37 +459,159 @@
+	// Sorts given arr according to DOM position.
+	//
+	// @param {CKEDITOR.dom.node[]} arr
+	TableSelection.prototype._arraySortByDOMOrder = function( arr ) {
```
It looks like I have lost it's usage during cleanups. It was introduced specifically for `insertColumn` call, as it [just appends created `td` elements into `selectedCells`](https://github.com/ckeditor/ckeditor-dev/blob/501b8b2/plugins/tableselection/plugin.js#L586), so this will append `n` columns from each row.

However I think it makes a good idea to put it in `_reset`. It's not a super performance-intensive function so that's going to be fine.

Comandeer replied:
Yes, I meant `setSelectedCells`.

mlewand commented on plugins/tableselection/plugin.js:
```diff
@@ -538,57 +754,43 @@
-			evt.data.dataValue = tmpContainer.getHtml();
-
-			selectCellContents( selectedCells[ 0 ] );
+		// In case of mixed content or non table content just insert it to a first cell, erasing the others.
```
Well this comments goes a bit outside of the scope of what table selection plugin is doing. I'll update it accordingly.

mlewand commented:
The bug is caused by [insertColumn](https://github.com/ckeditor/ckeditor-dev/blob/37583ef/plugins/tableselection/plugin.js#L590) as it returns cells in the whole column (as one might expect), while this logic assumes that only selected cells would be returned.

See a following selection: 
![image](https://user-images.githubusercontent.com/5353898/29126442-a9dd3118-7d1e-11e7-85bd-1a641ac57a5f.png)

In this case it adds to the `selectedCell` also highlighted "dummy cell".

I'll fix this bug shortly.

mlewand commented:
Fixed the bug caught above, in addition I played a bit with `rowspan` and added an improvement in 2a6f7ecee51a81f95f338eb83805a4c7b9a4f8d7. I was able to improve it, but it's not a priority to fix it completely (extracted to #754).
